doc/MiniLanguageTutorial.md

```diff
@@ -12,63 +10,59 @@ A Mini program is written as a set of source code files.  You can write separate
->
-> The optional `impure` modifier specifies that the function is impure, as defined below.
->
-> The compiler does not check whether the type signature declared here matches the type signature of the actual implementation of the function elsewhere.  If the type signatures are different, this might lead to a runtime error.
->
-> [Potential change: we could add the word unsafe to this somehow. Then we could say that every construction that can lead to an undetected type error has the word `unsafe` in its name.]
->
-> [Potential improvement: allow interfaces to be specified in separate files, and allow a code file to say that it implements or uses a particular interface. This would make it easier to maintain consistency, and easier for the compiler to check consistency..]
+> This declares a constant with name *ident* and value *unsignedInt*. This constant is only usable in the file it is defined.
 
-### Non-import declarations
+### Body declarations
 
 `type` *name* = *type*
 
 > This declares a type alias, allowing name to be used as a synonym for the specified type.  (Note that in the current syntax there is not a semicolon at the end.)
 
-var *name* : *type* ;
+`var` *name* : *type* ;
 
 > This declares a global variable. If type is an atomic type, the variable will be initialized to the zero value for that type. Otherwise the variable will be uninitialized. Reading an uninitialized variable before initializing it will cause undefined behavior.
 
-`[public] [impure] func` *name* ( *argname1: type1, argname2: type2, ...* ) [-> *returntype] codeblock*
+`#[` *Attributes* `]`? [ `view` | `write` | `public` ]* `func` *name* ( *argname1: type1, argname2: type2, ...* ) [-> *returntype*] *codeblock*
 
-`[public] [impure] func` *name* ( *argname1: type1, argname2: type2, ...* ) `noreturn` codeblock*
+`#[` *Attributes* `]`? [ `view` | `write` | `public` ]* `func` *name* ( *argname1: type1, argname2: type2, ...* ) `noreturn` codeblock*
 
 > This declares a function and provides its code.
 >
-> The `public` modifier is optional.  It indicates that the function can be called by code outside this source code file. Non-public functions cannot be called directly by outside code.  (However, pointers to non-public functions can be passed to outside code, and this would allow the pointed-to function to be called by outside code.)
+> The `public` modifier indicates that the function can be called by code outside this source code file. Non-public functions cannot be called directly by outside code.  (However, pointers to non-public functions can be passed to outside code, and this would allow the pointed-to function to be called by outside code.)
 >
-> The `impure` modifier is optional. It indicates that the function is impure, meaning that it might access global variables or call other impure functions.
+> The `view` modifier indicates that the function might read global variables or call other read functions.
```

Comment:
`view` & `write` always read or modify global state. We should probably have a description of what global state is, as well as indicate that you get a warning if a function is marked as more impure than it actually is.

Reply:
If you only get a warning if a function is marked more impure than it really is, wouldn't that mean that you cannot guarantee that it will always read or modify global state? Moreover, I believe because assignability allows non-`view` and non-`write` functions to be assigned to corresponding `view` and `write` types, this can also occur without a warning appearing. It is a good idea to explain what global state is, I'll add this.